[team-04] FE 3주차 IssueTracker

bread1022 opened the pull request:
안녕하세요 Autumn 🍂 
[워터](https://github.com/yesolyo)&[나니](https://github.com/bread1022)입니다! 이번 3주차에는 이슈목록페이지를 벗어나 이슈작성, 상세, 로그인 페이지를 구현하였고 지난번에 남겨주신 리뷰를 토대로 이슈목록페이지에는 리팩토링도 적용하고 상태와 기능을 적용하면서 고도화하기위해 많은 고민을 했던 한주였습니다! 아무래도 상태 적용이 처음이라 아직 뒤죽박죽하고, 설계도 어설프지만 최대한 구현해보았어용.. 이번주도 잘 부탁드립니다 🫠

## **🟢 구현내용**
![이슈트래커](https://github.com/codesquad-members-2023/issue-tracker/assets/107349637/4f1f9db9-6055-4171-b28a-b2c75d252d36)
![ㅎ](https://github.com/codesquad-members-2023/issue-tracker/assets/107349637/82fb081d-5ae5-42e7-b862-0d44e21afb19)


- [페이지별 컴포넌트 구조](https://www.figma.com/file/eJwjFCLzbmN3fphIYAsVYR/%EC%9D%B4%EC%8A%88%ED%8A%B8%EB%9E%98%EC%BB%A4-%ED%99%94%EB%A9%B4-%EA%B5%AC%EC%A1%B0%EB%8F%84?type=whiteboard&node-id=0-1&t=FmZrKqt5DqG7QBqv-0) 설계

- closed: https://github.com/codesquad-2023-group04/issue-tracker/issues/65 - 버튼 컴포넌트 스타일을 타입별로 변경하였습니다.
- closed: https://github.com/codesquad-2023-group04/issue-tracker/issues/69 - 이슈 목록 체크 박스를 클릭하여 이슈개수를 카운트할 수 있게 상태를 적용하였습니다. (이슈를 열고, 닫는 기능은 추가예정입니다)
- closed: https://github.com/codesquad-2023-group04/issue-tracker/issues/76,  
https://github.com/codesquad-2023-group04/issue-tracker/issues/77 - text input, text area 컴포넌트를 이슈작성페이지와 상세페이지에서 범용적으로 사용할 수 있게 구현하였습니다.
- closed: https://github.com/codesquad-2023-group04/issue-tracker/issues/87 - 로그인 페이지 레이아웃 구현을 완료하였고 Back-end 와 함께 Auth 연동하고자 노력중입니다.
- closed: https://github.com/codesquad-2023-group04/issue-tracker/issues/91 - Back-end에서 제공되는 api 명세에 맞춰 mock data 를 수정하였습니다.
- closed: https://github.com/codesquad-2023-group04/issue-tracker/issues/95 - text area 컴포넌트를 활용하여 comment element 컴포넌트 생성하였습니다.
- closed: https://github.com/codesquad-2023-group04/issue-tracker/issues/96 - issueDetail 페이지의 header 를 생성하였고 comment 작성 기능 추가 예정입니다.

## **🧩 고민과 해결과정**

#### 나니의 고민

- 드롭다운탭을 클릭할때마다 해당 탭의 api 를 fetch 하게끔 구현했는데 유저가 탭을 단순히 열고 닫았을 뿐인데 계속 패치하고 리렌더가 발생하는 것같습니다. 전후를 비교했을 때 데이터의 변동이 없고, 필요없는 리렌더를 방지하고싶은데 `useMemo`, `React.memo`, `useCallback` 등등 어떤 훅을(우선 제가 아는 건 이 세개뿐...), 어디에 적용해야할지 아직 감이 안와서 최적화를 적용하지 못했어요!
- 상태의 흐름을 디자인하고 설계하고 코드를 작성하기가 너무 어려워요...,,


#### 워터의 고민

- 테스트 코드를 짜고 싶은데, 현업에서 테스트 코드 어떤 툴을 사용하고 컴포넌트 단위로 코드를 짜는지 궁금합니다
- 처음 설계때부터 테스트 코드를 고려하면서 작성하는 지도 궁금합니다
- 백엔드에서 아직 api가 나오지 않았을때 post 보내면서 실시간으로 반영되는지 확인할 수 있는 방법이 있을까요?? 
- 새로운 이슈 작성페이지에서 `useState`로 최상단에서 관리하고 props로 함수를 컴포넌트 별로 전달해주고 (사이드 바, 제목, 코멘트) value를 완료 버튼으로 내려서 post를 보내려고 하는데 이런 식으로 하는게 맞는지 궁금합니다!

## **🗣️ 부가 설명, 리뷰 포인트**
- 지난번 리뷰에 대한 답변으로
`useContext`, `useReduer` 를 사용하게 된 이유는 상태와 컴포넌트에 필요한 옵션을 모두 props으로 전달해보니 중첩되어있는 컴포넌트에서 사용하지 않는 props 이어도 계속 전달 전달하는 로직을 조금 줄여보고 싶어서 `useContext` 를 사용했어요!
reducer 는 이전 미션(뉴스스탠드)에서 바닐라로 서툴게 사용해보고,, 리액트 `useReducer` hook을 강의로 접하면서 다시 한번 시도해보고 싶었습니다. ( 마침 크롱도 `useState`, `useReducer`, `useContext` 이 세가지 조합을 활용해 보면 좋다고 말씀해 주신 것도 한 몫했어요:) ) 또 이번 이슈트래커에서 관리해야 하는 공통된 상태(open, close, author 등등)를 여러 컴포넌트들이 공유하고 또 같은 액션을 하고 있다고 생각이 되어서 리듀서를 사용하면 조금 더 로직이 깔끔해지지 않을까(?)싶어서 적용해보려고.. 노력중입니다 🥲
- 삼주차 리뷰포인트는 우선 저희가 `"상태관리가 처음이라는 점"`인데요, 앞으로 상태관리를 이런식으로 해도 좋을지 ? 더 나은 방법이 있는지 ! 입니다 🥹

bread1022 committed:
REFACTOR: Header 컴포넌트분리

- Header 컴포넌트를 페이지와 분리하여 로그인시 해당 유저의 정보를 받아 표시하였습니다.
- Profile 컴포넌트의 범용성을 높이기 위해 width, height props를 추가하였고, default 값을 설정하였습니다. (isLarge면 32, 아니면 20)

bread1022 committed:
REFACTOR: 드롭다운 컴포넌트 범용성 향상, 외부 영역 클릭 이벤트 적용

- 하나의 Dropdown 컴포넌트로 항목이 다른 필터탭도 사용할 수 있게끔 로직을 변경하였습니다.
- 필터탭 외부 클릭시 패널을 닫히게끔 이벤트를 적용하였습니다.

bread1022 committed:
MERGE: #23-add-state-issueList

Fe feature/#23 add state issue list

bread1022 committed:
REFACTOR: 스타일컴포넌트 하단으로 위치 변경

- 스타일 컴포넌트 컨벤션에 맞게 수정

bread1022 committed:
MERGE: 리뷰 반영하여 코드 수정 진행

yesolyo committed:
ADD: mock db 변경 완료

-api가 변경되어 mock db를 변경 완료하였습니다.

yesolyo committed:
ADD: comment db 설정 완료

- comment db를 포함시켜야해서 추가하였습니다.

yesolyo committed:
MERGE: #64 mockData 변경 완료

Fe sprint3

bread1022 committed:
CHORE: mock data 변경에 따른 로직 수정

bread1022 committed:
MERGE: mock data, 로직 변경

CHORE: mock data 변경에 따른 로직 수정

yesolyo committed:
FEAT: 버튼 컴포넌트 스타일 변경 (#65)

- 버튼 컴포넌트 스타일을 통일성의 필요성을 느꼈습니다.
- 통일성을 고려해서 컴포넌트 내부 안에서 스타일을 세분화하였습니다.

yesolyo committed:
MERGE: #65-change-button-style

FEAT: 버튼 컴포넌트 스타일 변경 (#65)

bread1022 committed:
ADD: mock data 변경에 따른 드롭다운탭 로직 변경

- mock data 변경에 따른 상태변경을 위해 reducer 를 생성하였습니다.
- mock data 변경에 따른 내부 컴포넌트의 key 값을 id 로 변경하였습니다.

bread1022 committed:
MERGE: 버튼 리팩토링 적용하여 병합

bread1022 committed:
ADD: 버튼 로직에 맞게 코드 수정

bread1022 committed:
MERGE: #67-add-dropdownTabs-filter

#67 add dropdown tabs filter

yesolyo committed:
FEAT: 사이드바 드롭다운 탭 구현 (#79)

- 사이드바 드롭 다운 탭을 구현하였습니다.
- z-index 문제가 발생하였고, 협업 과정 중에서 셀렉 문제가 발생하였습니다.
- 추후 협업을 통해서 문제해결을 진행할 예정입니다.

yesolyo committed:
FEAT: text input 클릭 시, label이 위로 올라가는 상태 적용

- text input을 클릭 시, label이 위로 올라가는 상태를 적용하였습니다.
- 기존의 text input은 icon도 있는지 여부를 확인하는 형태였지만 문제점이 서로 기능이 다르다고 생각하여 변경하였습니다.
- 그래서 icon text input인 filter bar를 변경하였습니다.

bread1022 committed:
REFACTOR: 드롭다운탭 범용성 향상

- 이슈목록조회와 새이슈작성페이지에서 사용가능한 드롭다운탭으로 로직을 수정하였습니다.

bread1022 committed:
MERGE: Dropdown 리팩토링 병합

- REFACTOR: 드롭다운탭 범용성 향상
- state 주입

yesolyo committed:
FEAT: text area 컴포넌트 상태 설정 완료 (#76)

- 클릭 시 label이 변동되게 설정 완료
- 클릭 후 입력을 하고 다른 데를 클릭해도 label이 변동된 상태가 된다.
- 만약 입력을 안했을 시, label이 원래 모습을 보여준다.

yesolyo committed:
Merge branch 'FE-sprint3' into FE-feature/#76-change-text-input

yesolyo committed:
MERGE: #76-change-text-input

Fe feature/#76 change text input

bread1022 committed:
CHORE: Icon label 수정에 따른 로직 변경

- 디테일 수정 (아이콘 크기, 버튼 크기 변경)

bread1022 committed:
MERGE: 아이콘 크기 변경에 따른 수정 사항 반영

CHORE: Icon label 수정에 따른 로직 변경

yesolyo committed:
FEAT: 로그인 페이지 구현 (#87)

- 로그인 페이지 레이아웃을 구현하였습니다.
- 아이디와 비밀번호 입력 시 버튼이 활성화됩니다.

yesolyo committed:
Merge branch 'FE-sprint3' into FE-sprint3

yesolyo committed:
MERGE: #88 FE-sprint3

FEAT: 로그인 페이지 구현 (#87)

yesolyo committed:
ADD: 로그인 페이지 이슈 트레커 header 표시 (#87)

- 로그인 페이지 이슈 트래커 로고가 없어서 로고 제작하였습니다.

bread1022 committed:
ADD: IssueList, FilterState 전역상태관리 (#67)

- filter 상태에 따라 이슈리스트를 관리하기 위해 Provider를 사용하였습니다

bread1022 committed:
ADD: checkbox state 관리

- 이슈목록컨테이너에서 이슈상태관리를 적용하였습니다.

yesolyo committed:
FEAT: mock 데이터 생성

yesolyo committed:
DEL: 쓸모없는 코드와 파일을 삭제하였습니다.

- 코드의 재구성이 필요하여 쓸모없는 코드와 파일을 삭제합니다.

yesolyo committed:
MERGE: #91-add-mock

Fe feature/#91 add mock

bread1022 committed:
FEAT: 체크박스 상태 적용 (#69)

- 체크박스 컴포넌트를 수정하고 이슈 목록 조회페이지의 컨테이너에서 체크박스 컨텍스트를 생성하여 전역으로 상태관리하였습니다.
- 체크박스 리듀서를 생성하여 이슈 선택, 해제 & 전체이슈 선택, 해제 기능을 구현하였습니다.

bread1022 committed:
Merge branch 'FE-sprint3' into FE-feature/#67-add-state-checkbox

bread1022 committed:
MERGE: 체크박스 상태 적용 후 병합

#67 add state checkbox

bread1022 committed:
CHORE: DropdownTab 로직 수정 (#67)

- 사이드바, 탭버튼 에서 공통으로 사용하는 로직을 분리하였습니다.
- 불필요한 코드를 삭제하였습니다.

yesolyo committed:
FEAT: issue Detail 헤더 구현 완료 (#96)

- information tag를 제작하려고 하였으나, 기존 label tag가 있어서 이 tag로 재사용하였습니다.
- mock data를 제작하여 데이터를 이용해서 헤더를 구성하였습니다.

bread1022 committed:
IMPROVE: Dropdown tab mousedown 이벤트적용 (#67)

- 드롭다운탭을 클릭했을때 각 탭의 정보를 가진 api에 fetch하는 로직으로 변경하였습니다.

bread1022 committed:
FIX: 드롭다운탭 onMouseDown 이벤트 적용시 옵션 선택 안되는 버그 해결 (#67)

- onClick 으로 변경하고, if 문으로 예외처리 진행하였습니다.

yesolyo committed:
FEAT: comment 컴포넌트 구현 (#95)

- comment 컴포넌트를 구현하여서, comment 수만큼 컴포넌트가 생성되게 구현하였습니다.

bread1022 committed:
MERGE: 드롭다운탭 mousedown 이벤트 적용 후 병합

#67 improve dropdown

yesolyo committed:
MERGE: #95-add-comment-elements

Fe feature/#95 add comment elements

yesolyo committed:
CHORE: 코드 수정 완료

- header 클릭 시 이슈리스트로 이동하게 설정하였습니다.
- 기본 코드들을 간결하게 하기 위해서 수정하였습니다.

yesolyo committed:
MERGE: /#97-chore-file

CHORE: 코드 수정 완료

yesolyo committed:
FEAT: 아이디와 비번을 입력 시 user 있는 지 확인 (#108)

- 아이디와 비번을 입력하면 user data에 있는지 확인한다.
- 있으면 이슈리스트 페이지로 넘어간다.
- 없으면 비밀번호에 포커스 된다.

yesolyo committed:
ADD: 로그인 클릭 시 사용자가 없을 경우 오류 표시

- 로그인 클릭 시 사용자가 없을 경우 focus되며 테두리가 빨간 색으로 표시된다.

bread1022 committed:
IMPROVE: Sidebar 정보 설정 기능 구현 (#79)

- 사이드바에서 담당자, 레이블, 마일스톤을 지정하면 사이드바에 즉시 표시되게 하였습니다.
- 필요없는 주석과 변수명을 약간 수정하였습니다.

bread1022 committed:
MERGE: #79-dropdowntab-sidebar-improve

IMPROVE: Sidebar 정보 설정 기능 구현 (#79)

hongbiii commented on FE/src/assets/Icon.js:
```diff
@@ -1,19 +1,18 @@
       width={width}
-      height={height}
-      stroke={stroke}
+      height={height || width}
```
p3)
디폴트값을 설정하는 것은 기존에 해주셨던 것처럼 default parameter를 사용하는 게 더 괜찮아보입니다!

width 만 넘겨줬을 때 height 에 동일한 값을 적용한다 (정사각형처럼?) 라는 의도가 담긴 걸까요??
이 Icon 컴포넌트를 사용할 때 width, height 값을 동일하게 설정하고 싶을 때 width 만 설정해주면 되겠다! 라고 생각이 이어지지는 않는 것 같아요.

bread1022 replied:
width, height 를 명시해주지 않으면 정사각형으로 동작하게끔 구현하고싶었는데 어텀의견에 힌트를 얻어서 다시 수정해봤어요! 감사합니다
```js
const Icon = ({
  iconType,
  width = 16,
  height = width,
  fill = colors.gray50,
  onClick
}) => {
  const IconSvg = icons[iconType];
  return (
    <IconSvg width={width} height={height} fill={fill} onClick={onClick} />
  );
};
```

hongbiii commented on FE/src/utils/timeElapsed.js:
```diff
@@ -27,8 +27,15 @@ export const getTimeElapsed = (date) => {
+  } else if (timeDifference < timeOptions.weeks * 4) {
     timeUnit = 'weeks';
+  } else {
+    return new Date(date).toLocaleDateString('ko-KR', {
+      year: 'numeric',
+      month: 'numeric',
+      day: 'numeric',
+      weekday: 'long'
+    });
```
toLocaleDateString 메서드에 이런 옵션이 있었군요~ 배워갑니다. ㅎㅎ

(참고)
날짜를 다룰 때는 라이브러리를 많이 활용하곤 하는데요.
라이브러리를 사용하면 훨씬 쉽고 간단하게 포맷팅을 할 수 있습니다.
라이브러리 코드를 까보진 않았지만 내부적으로 toLocaleDateString 메서드의 옵션이 적극 사용되지 않을까 싶네요!
toLocaleDateString 으로 직접 포맷팅을 해보셨으니 라이브러리를 사용하게 되시면 라이브러리에서 어떤 일을 해주는지 빠르게 이해하실 것 같아요 👍👍

hoongding replied:
저도 한수요..!

hongbiii commented on FE/src/App.js:
```diff
@@ -1,25 +1,37 @@
-          {/* TODO : user 로그인 하면 헤더 고정 시키기 <Header /> */}
+          {/* {login.isLogin && <Header />} */}
+          {<Header />}
           <Routes>
-            <Route path="/" element={<IssueList />} />
+            <Route path="/" element={<Login />} />
```
(궁금) 혹시 로그인페이지에서도 헤더가 표시되지는 않나요?

bread1022 replied:
맞습니다~~~~~ 로그인 처리가 완료되면, 헤더를 표시하게끔 `login.isLogin && <Header />` 하려고 했는데 로그인 설정중이라 헤더를 띄워놨어요! 😲🫣😱

hongbiii replied:
react router dom 을 사용하시는 김에 layout을 적용해보는 건 어떨까요?
https://reactrouter.com/en/main/start/concepts#layout-routes

hongbiii commented on FE/src/App.js:
```diff
@@ -1,25 +1,37 @@
+            <Route path="/auth" element={<Auth />} />
+            <Route path="/issues" element={<IssueList />} />
             <Route path="/newIssue" element={<NewIssue />} />
+            <Route path="/issueDetail/:id" element={<IssueDetail />} />
```
p4)
뒤에 id가 붙기 때문에 detail이라는 단어는 없어도 괜찮습니다. 😀
```suggestion
            <Route path="/issues/:id" element={<IssueDetail />} />
```

bread1022 replied:
이슈 작성하면 바로 해당 아이디 페이지로 넘어가니까 detail 페이지가 따로 필요하지 않겠네요 ! 참고해서 로직 수정해보겠습니다 :)

hongbiii commented on FE/src/components/Header.js:
```diff
@@ -1,42 +1,47 @@
-`;
+export const Header = () => {
+  const navigate = useNavigate();
+  const [user, setUser] = useState('');
```
state의 디폴트값을 설정해줄 때는 해당 state의 타입에 맞게 설정하거나 없다면 명시적으로 null로 설정해주는 것이 좋습니다!
`''` 빈 스트링을 보니 user라는 상태의 타입이 string일 거라고 예상하게 되는데 실제로는 아래와 같이 타입설정이 되어 있네요~

```ts
  {
    "id": number,
    "name": string,
    "profileUrl": string
  }
```

bread1022 replied:
디폴트는 null 로 하는 게 좋군요.. 그런 디테일까지 맞춰야하는지 몰랐네요ㅠㅠ 참고해서 수정해볼게요 ! 감사함당 ෆ

bread1022 committed:
IMPROVE: 드롭다운탭 옵션 선택 기능 완료 (#67)

- 드롭다운탭에서 옵션을 선택하고, 재선택하면 해제되는 기능을 구현하였습니다.
- 필터바, 이슈목록 헤더 탭버튼, 사이드바 에서 사용할 수 있게 코드를 수정하였습니다.

bread1022 committed:
MERGE: #79-dropdowntab-sidebar-improve

- IMPROVE: 드롭다운탭 옵션 선택 기능 완료 (#67)
- 불필요한 코드 수정

hongbiii commented on FE/src/components/PageHeader.js:
```diff
@@ -0,0 +1,31 @@
+      {rigthChild && <div>{rigthChild}</div>}
+    </MyPageHeader>
+  );
+};
```
p1)
공통 부분을 최대한 재사용하기 위한 고민의 흔적이 보이네요!

두 가지 관점에서 생각해볼 수 있을 것 같아요.
1. 진짜 재사용이 가능한가?
2. props와 로직이 적절한가?

hongbiii replied:
> 1. 진짜 재사용이 가능한가?

PageHeader 컴포넌트가 어디에서 사용됐는지 보니,
이슈 작성 페이지와 이슈 상세 페이지에서 사용됐네요!

재사용이 가능한지 판단할 때 해당 요소들이 같은 위계에 있는지를 생각해보시면 도움이 될 것 같아요.

페이지 헤더(사실 페이지 타이틀이 더 정확한 표현 같기도?)는 어떤 느낌이냐면
[배민문방구](https://brandstore.baemin.com/) 이 사이트의 GNB를 클릭했을 때 나오는 페이지 제목들! 그런 느낌이에요.
<img width="1270" alt="image" src="https://github.com/codesquad-members-2023/issue-tracker/assets/125108533/631fc959-c754-47d8-a5fa-85fc4f7eb2ac">
<img width="1259" alt="image" src="https://github.com/codesquad-members-2023/issue-tracker/assets/125108533/179d6e9f-f525-4a87-adba-d3106bb61cae">

우리 이슈 트래커에서 이슈 작성 페이지의 "새로운 이슈 작성" 은 페이지 타이틀이라고 볼 수 있겠어요.
그럼 이슈 상세 페이지의 제목은요?!
이 제목은 페이지 타이틀보다는 특정 이슈의 제목 이라고 보는 게 맞는 것 같아요.
배민문방구로 다시 예를 들어보자면 상품을 클릭해 들어갔을 때 보이는 상품의 이름과 같은 위계예요.
페이지 타이틀보다 위계가 낮다고 볼 수 있습니다!


<img width="1286" alt="image" src="https://github.com/codesquad-members-2023/issue-tracker/assets/125108533/f2bf22fd-607e-4fb7-8048-6753422e7fe2">


물론 페이지타이틀과 이슈 제목의 스타일이나 컴포넌트 구성이 완전히 똑같거나 약간의 차이가 있다면 컴포넌트를 재사용할 수도 있는데요!
그럴 때는 조금 더 추상화해서 정말 일반적으로 사용할 수 있는 컴포넌트로 만들어 사용해야 합니다.
작성해주신 코드는 PageHeader 라는, 위계를 나타내는 이름을 가졌지만 내부 구현은 OOOChild 로 추상화가 되어있어 혼선이 생기는 것 같아요.

hongbiii replied:
>2. props와 로직이 적절한가?

우선 left, middle, right 로 구분을 하는 게 이슈 상세페이지를 그리는 상황에서는 적절하지 않은 것 같아요.
우측 버튼 2개를 묶어서 right로 보신 것처럼 이슈제목-이슈번호도 한 덩어리로 보는 게 맞습니다!

개인적으로는 이슈 상세페이지의 제목 부분은 별도의 컴포넌트로 구성하고 더 명확한 props 이름을 사용하는 게 더 적절하다고 생각해요.

지금처럼 PageHeader 하나로 구현을 하고 싶으시다면 위와 같은 이유로 left, right 정도로 구분하는 게 좋을 것 같구요.
(이때는 컴포넌트 이름이 좀 바뀌어야 이해하기 쉬울 듯 😅)
또다른 이유로는, 제목 편집 버튼을 클릭하면 제목 부분이 input 으로 바뀌어야 하는데 middle이 계속 존재한다면 구현이 복잡해질 것 같아요.
그리고 left나 right에 어떤 요소가 올지 예측할 수 없기 때문에 스타일을 PageHeader 내부에서 구현하기보다는 외부에서 주입해주는 것이 좋겠습니다!
리액트에서는 props로 리액트 컴포넌트를 넘겨줄 수도 있어서요,
스타일까지 완성된 컴포넌트를 left, right에 넘겨주면 됩니다. 😀

hongbiii replied:
사실 그렇게 복잡하거나 심각한 내용은 아닌데 글로 남겨서 장황해졌네요 🥲🥲🥲;;;
몇 번 구현을 하다보면 어떻게 나눠야할지 감이 오실 거라 혹시나 너무 걱정은 안 하셔도 돼요!!

hongbiii commented on FE/src/components/button/Button.js:
```diff
@@ -1,93 +1,51 @@
-  };
-
-  const MyButton = buttonTypes[type];
+  const btnSize = btnSizes[size];
```
좋습니다~ 👍👍👍
요렇게 내부 구현(버튼 width, height의 정확한 px 값)을 컴포넌트 안으로 숨기고 (`캡슐화`)
외부(props)에는 옵션(l, m, s, xs)만 노출하면 (`인터페이스`)
변경에 유연해진다는 장점이 있어요!

여기에 더불어서
디자이너와 소통을 할 때도 
> 버튼 사이즈는 width 320px, height 56px 이에요.

보다는

> 버튼 사이즈는 large 입니다.

라고 커뮤니케이션 하는 게 더 편하기도 하구요!

hongbiii commented on FE/src/components/button/Button.js:
```diff
@@ -1,93 +1,51 @@
+      btnSize={btnSize}
+      btnColor={btnColor}
       onClick={onClick}
+      type="button"
```
👍

hongbiii commented on FE/src/pages/Login.js:
```diff
@@ -0,0 +1,109 @@
+          setValue={login.setValue}
+        />
+      ))}
+      <Button {...idLoginBtn} />
+      <Button {...registerBtn} />
+    </MyLogin>
+  );
+};
+
+const MyLogin = styled.form`
```
form 요소를 사용하셨군요!
지금 Button 컴포넌트 내부에 type="button" 으로 고정되어 있는데
기본 타입은 button이면서 다른 타입도 받을 수 있도록 살짝 고쳐보면 어떨까요?
그러면 버튼을 클릭해서 submit 이벤트를 발생시킬 수 있을 것 같아요

hongbiii commented on FE/src/utils/fetch.js:
```diff
@@ -0,0 +1,34 @@
+  return resData;
+};
+
+export const fetchAll = async (...url) => {
+  const response = [...url].map((url) => fetch(url));
```
[Rest paremeter](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/rest_parameters#description)는 배열에 담기기 때문에 얕은 복사를 해주지 않아도 돼요!

fetch 대신 위에 만들어둔 fetchData를 사용할 수 있을까요??

bread1022 replied:
사용이..되네요 ! 제가 저번 미션부터 계속 사용하던 함수인데 덕분에 더 깔끔해졌어용 ㅜㅜ 🫶🏻

hongbiii commented on FE/src/utils/fetch.js:
```diff
@@ -0,0 +1,34 @@
+  return response;
+};
+
+export const fetchPost = async (url, body) => {
+  const options = {
+    method: 'POST',
+    headers: {
+      'Content-Type': 'application/json'
+    },
+    quries: { code: queryCode }
+  };
```
인자로 넣어준 body는 사용되지 않고 있고 quries 오타가 있는 듯 해요!
queryCode도 정의되어 있지 않은 변수 같아요
요 함수 다시 확인이 필요할 것 같슴당

hongbiii commented on FE/src/pages/IssueDetail.js:
```diff
@@ -0,0 +1,27 @@
+    initData();
+  }, []);
+  return (
+    <IssueDetailContext.Provider value={issueDetail}>
+      <IssueDetailHeader />
+      <IssueDetailContainer />
+    </IssueDetailContext.Provider>
```
p3)

본문에 남겨주신 것처럼 props drilling이 있을 때 context는 좋은 대안이 됩니다!
그치만 IssueDetailHeader 의 경우에는 1depth 인데 이때도 context 로 관리해야 할까? 하는 생각이 드네요!

작성해주신 코드에서는 페이지 컴포넌트 레벨에서 해당 페이지에 필요한 모든 데이터를 fetch 하고 있는데요,
이렇게 작성하게 되면 하나의 API에서 에러가 발생했을 때 화면 전체가 안 나오는 문제가 발생할 수 있습니다!
fetch 하는 단위를 작게 가져가면 페이지 내에서 에러가 발생한 컴포넌트만 부분적으로 에러 처리를 할 수 있고
props drilling도 막을 수 있기 때문에 이 부분 한 번 고려해보셨으면 좋겠어요 :)

hongbiii commented on FE/src/components/textForm/TextArea.js:
```diff
@@ -0,0 +1,135 @@
+        onBlur={() => setIsTextAreaFocus(false)}
+      />
+      <label className={value && 'filled'}>{label}</label>
+      <MyFileArea isFocus={isTextAreaFocus} fileSize={fileSize}>
+        <Button
+          size={'m'}
+          color={'ghostBlack'}
+          iconType={'paperclip'}
+          isIcon
+          buttonText={`파일 첨부하기`}
+          isLeftPosition
+        />
+        <div>띄어쓰기 포함 {value.length}자</div>
+      </MyFileArea>
```
파일 첨부의 경우 [input type="file"](https://developer.mozilla.org/ko/docs/Web/HTML/Element/input/file)을 사용해서 구현하면 됩니당 😀
파일 쪽도 한 번 찍먹 해보시면 좋긴 한데 보통의 json 타입과는 다르게 신경써줘야 할 부분들이 있어서 우선순위는 낮게 보셔도 (지금은) 괜찮아요!
어차피 나중에 다 하게 되어있읍니다.. 껄껄

bread1022 committed:
REFACTOR: TabButton 수정 (#67)

- 마스터클래스 리뷰반영
- Filter Section에 TabButton 분리 필요성이 낮아보여 분리하지않는 방향으로 수정하였습니다.

yesolyo committed:
ADD: mock 데이터 생성

- 각 필터 별로 mock data 가 필요하여 생성하였습니다.
- 기준은 user id 1번을 기준으로 data를 생성하였습니다.

yesolyo committed:
Merge branch 'FE-sprint3' into FE-feature/#108-login-error

yesolyo committed:
MERGE: #108-login-error

Fe feature/#108 login error

hongbiii commented on FE/src/pages/NewIssue.js:
```diff
@@ -2,39 +2,52 @@ import React, { useEffect, useState } from 'react';
   // TODO : Fetch -> user 정보 필요
   const initData = async () => {
-    const response = await fetch('/issueList');
+    const response = await fetch('/issues');
```
새 이슈 페이지인데 왜 issues 데이터를 불러오고 있는지 궁금하네용 👀

bread1022 replied:
유저정보가 필요해서 가져왔던 것같은데, 추후에 로그인 기능이 적용되면 즉시 로직 수정해서 반영해보겠습니당

bread1022 committed:
MERGE: Text Input 수정 사항 병합 (#108)

hongbiii commented on FE/src/pages/NewIssue.js:
```diff
@@ -2,39 +2,52 @@ import React, { useEffect, useState } from 'react';
-      </NewIssuePage>
+      <MyNewIssuePage>
+        <PageHeader leftChild={'새로운 이슈 작성'} />
+        <NewIssueContainer
+          titleValue={title}
+          titleSetValue={setTitle}
+          commentValue={comment}
+          commentSetValue={setComment}
+          assigneeSetValue={setAssignee}
+          labelSetValue={setlabel}
+          milestoneSetValue={setMilestone}
+        />
```
p2)

NewIssueContainer 컴포넌트로 존재하지 않아도 될 것 같아요.
살짝 과하게 컴포넌트화가 된 느낌적인 느낌입니닷
작성 영역과 담당자/레이블/마일스톤 지정 영역이 서로 영향을 미치지 않기 때문에 하나의 컴포넌트로 묶어주지 않는 게 좋아보여서요!

그리고 추가로,
상태는 최대한 작은 컴포넌트에서 관리를 해보고 그게 불가능할 때 부모 컴포넌트로 한 단계씩 올려보는 것을 시도해보심 좋을 것 같아요.
지금은 페이지 단위(거의 최상위 컴포넌트)에서 모든 상태를 관리하려다 보니 props drilling이 발생하고 또 이걸 해결하기 위해 불필요하게 context가 많아지고 있어요.

hongbiii replied:
state와 context 관련해서 리팩토링이 진행되어야 memo 사용하신 게 적절한지 판단을 해볼 수 있을 것 같습니당 :)

hongbiii commented on FE/src/utils/filterQuery.js:
이 파일의 두 함수가 쪼끔 복잡해보이는데 아직 사용되는 곳이 없어서 다음 주에 다시 한 번 볼게요!!

hongbiii commented on FE/src/stores/filterStateReducer.js:
```diff
@@ -0,0 +1,28 @@
+    case 'RESET-FILTER': {
+      return initialFilterState;
+    }
+    case 'FILTER-ISSUES': {
+      return {
+        ...initialFilterState,
+        ...payload
+      };
+    }
```
현재 상태를 기반으로 상태 변경이 일어나야 해요!
```suggestion
    case 'FILTER-ISSUES': {
      return {
        ... state,
        ...payload
      };
    }
```

hongbiii reviewed (approved):
워터, 나니 안녕하세요! 이번 주도 열심히 작업하셨군요 🥳
첨부해주신 gif를 보니 기능 동작이 다 잘 되고 있네요~~ 👍
다음주부터는 실제 백엔드 API를 슬슬 붙여보시면 좋겠습니다!
아무리 목데이터를 잘 짜놨다 하더라도 실제 FE-BE 통합을 해보면 항상 삐걱거리거든요. ㅎㅎㅎ
백엔드 실행하는 법 여쭤보고 진행해보셔요 !!

파일이 많아서 제가 Dropdown 쪽은 확인하지 못했는데 다음주에 다시 보도록 하겠습니다!
얼추 대략적인 공통 컴포넌트들은 개발이 된 것 같아 보이네요 👍

티켓 관리, 문서는 항상 잘 하고 계셔서 감탄하고 있습니다. ㅎㅅㅎ
매주 새로운 학습거리가 나오는데 부지런히 적용해보고 계신 점도 대단하구요!!
(저는 다 못했었거든요 ㅋㅋㅋㅋㅋㅋㅋㅋ)

아참 이번주는 [코드리뷰 Pn룰](https://blog.banksalad.com/tech/banksalad-code-review-culture/#%EC%BB%A4%EB%AE%A4%EB%8B%88%EC%BC%80%EC%9D%B4%EC%85%98-%EB%B9%84%EC%9A%A9%EC%9D%84-%EC%A4%84%EC%9D%B4%EA%B8%B0-%EC%9C%84%ED%95%9C-pn-%EB%A3%B0)을 살짝 도입해봤어요.
p1에 가까울수록 적극 반영해주세요, p5로 갈수록 그냥 의견 이에요!

정말정말 고생 많으셨습니당!
남겨주신 질문은 코멘트 하나 더 남겨둘게요~!

hongbiii reviewed (approved):
워터, 나니 안녕하세요! 이번 주도 열심히 작업하셨군요 🥳
첨부해주신 gif를 보니 기능 동작이 다 잘 되고 있네요~~ 👍
다음주부터는 실제 백엔드 API를 슬슬 붙여보시면 좋겠습니다!
아무리 목데이터를 잘 짜놨다 하더라도 실제 FE-BE 통합을 해보면 항상 삐걱거리거든요. ㅎㅎㅎ
백엔드 실행하는 법 여쭤보고 진행해보셔요 !!

파일이 많아서 제가 Dropdown 쪽은 확인하지 못했는데 다음주에 다시 보도록 하겠습니다!
얼추 대략적인 공통 컴포넌트들은 개발이 된 것 같아 보이네요 👍

티켓 관리, 문서는 항상 잘 하고 계셔서 감탄하고 있습니다. ㅎㅅㅎ
매주 새로운 학습거리가 나오는데 부지런히 적용해보고 계신 점도 대단하구요!!
(저는 다 못했었거든요 ㅋㅋㅋㅋㅋㅋㅋㅋ)

정말정말 고생 많으셨습니당!
남겨주신 질문은 코멘트 하나 더 남겨둘게요~!

bread1022 committed:
CHORE: Text Input, Login 수정사항 반영 (#108)

- Text input focus 이벤트 수정에 따른 Login page 코드를 수정하였습니다.

bread1022 committed:
MERGE: #108 Login, Text input focus event 수정

#108 로그인페이지 포커스 이벤트 수정 사항 반영

hongbiii commented:
남겨주신 질문에 대한 제 생각을 남겨봅니다. 🙂

> 계속 패치하고 리렌더가 발생

제가 직접 실행해본 건 아니라 정확히 어떤 부분이 리렌더링 되는지는 확인해보지 못했는데요. 😅
최적화를 하려는 목적을 정의해보고 그에 맞는 방법을 적용해보시면 좋을 것 같아요!

- 불필요하게 깜빡임이 발생하는 것이 유저에게 불편함을 줘서?
- API 콜을 너무 많이 하게 돼서?
- 리렌더링이 너무 느리게 돼서?

저도 최적화 부분이 정말 어렵게 느껴졌었는데
다른 검색 하지 마시구 일단 [리액트 beta docs](https://react.dev/reference/react/useCallback) 이 문서 쭉 한 번 읽어보세요!!
Troubleshooting 전까지만 읽어보시면 어떤 상황에 useCallback, useMemo 를 써야 하는지 조금 감이 오실 거예요.

그리고 고민하시는 부분에 힌트가 될까 하여 또 다른 키워드도 드릴게요! ....... ˗ˋˏ `API cache` ˎˊ˗

---

> 상태의 흐름을 디자인하고 설계

ㅎㅎㅎ 처음 해보는 거라 그런 것 같아요.
리뷰에도 남겼지만 지금 API 콜이 페이지 컴포넌트 단위에서 이루어지고 있어서 상태를 다루는 게 복잡해지지 않았나 싶네요!
최대한 작은 컴포넌트 단위에서 상태를 처리해보고, 그게 안 되면 상태를 한 단계씩 끌어올려보고, 여기서 props drilling에 대한 불편함이 생기면 그때 context 를 도입해보셔용

---

> 현업에서 테스트 코드 어떤 툴을 사용하고 컴포넌트 단위로 코드를 짜는지
> 처음 설계때부터 테스트 코드를 고려하면서 작성하는 지

저희는 주로 jest 를 사용하고, vitest를 사용하는 팀도 있어요!
저는 유틸함수에 대해서만 테스트코드를 작성하고 있어요.
컴포넌트를 테스트해야 하는 필요성은 아직 느껴지지 않아서요,, (테스트코드 잘 모름...... 🤧)
테스트코드를 많이 작성하는 편은 아니라서 설계 때부터 테스트코드를 고려하진 않구요!
다만 유틸함수 테스트코드를 작성할 수 있도록 최대한 컴포넌트 내부 상태에 의존적이지 않게 순수함수로 만들어서 컴포넌트 바깥에 함수 선언을 하고 있습니다.

테스트코드에 관심이 있으시다면 테스트코드에 대해 많은 고민을 하고 계신 저희 팀원 분을 한 번 연결시켜드리겠습니당ㅋㅋㅋㅋ

---

> 백엔드에서 아직 api가 나오지 않았을때 post 보내면서 실시간으로 반영되는지 확인할 수 있는 방법이 있을까요??

실제로 반영이 되는지 여부라면 직접 DB 구성을 하셔야 할 것 같은데요
들어가는 공수 대비 큰 효과는 없어서 (API 개발 완료되면 필요없기도 하니까) 비추합니다!!
목서버 구성하고 네트워크 요청이 정상적으로 가는지만 확인해보셔도 충분하다고 생각해요. 😙
개발하실 때 어떻게 하고 계신지를 몰라서.. 혹시 개발자도구의 네트워크 탭을 잘 활용하고 계신지, 안 보고 계셨다면 이참에 한 번 보면서 개발하는 것을 추천드립니다!

---

> 새로운 이슈 작성페이지에서 useState로 최상단에서 관리하고 props로 함수를 컴포넌트 별로 전달해주고 (사이드 바, 제목, 코멘트) value를 완료 버튼으로 내려서 post를 보내려고 하는데 이런 식으로 하는게 맞는지 궁금합니다!
> 상태관리를 이런식으로 해도 좋을지 ? 더 나은 방법이 있는지

새 이슈 작성 페이지에서 page 컴포넌트에서 모든 state를 정의하고 자식 컴포넌트에게 내려주고 있는데,
NewIssueContainer.js 에서 리턴하는 부분을 살펴보면 NewIssueContent, SideBar 에 겹치는 props가 없기 때문에
상태를 부모컴포넌트(page 컴포넌트)에서 관리할 필요가 없어 보여요.

```jsx
      <NewIssueContent
        titleValue={titleValue}
        titleSetValue={titleSetValue}
        commentValue={commentValue}
        commentSetValue={commentSetValue}
      />
      <SideBar
        assigneeSetValue={assigneeSetValue}
        labelSetValue={labelSetValue}
        milestoneSetValue={milestoneSetValue}
      />
```

그치만?!
이렇게 상위 컴포넌트에서 상태 관리를 하신 것이 form을 다루기 위한 큰 그림이었다면 괜찮은 시도 같습니다!

˗ˋˏ `controlled/uncontrolled component` ˎˊ˗  살짝 키워드를 놓아두고 갑니다~~~!
참고로 작성해주신 새이슈작성 페이지 코드처럼 form을 state로 다루는 것을 controlled component 라고 하는데요,
react hook form 이라는 라이브러리에 적용되어 있는 방식이기도 합니다!
